File: cyberShop/src/main/java/com/company/cyberShop/dto/Product.java
package com.company.cyberShop.dto;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

import javax.persistence.*;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Pattern;
import java.util.Objects;

@Entity
@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
@Table(name="products")
public class Product {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;
    @NotEmpty(message= "product name cannot be empty")
    private String name;
    @NotNull(message= "price cannot be empty")
    private Float price;
    @NotEmpty
    @Pattern(regexp = "domestic|imported", message = "Shipping status must be domestic or imported")
    private String shipping;
    @NotEmpty
    @Pattern(regexp = "Books|Food|Medical|Music|Luxury Items|Clothes")
    private String category;
    private String image;

    public String getImage() {
        return image;
    }

    public void setImage(String image) {
        this.image = image;
    }

    public Integer getId() {
        return id;
    }



    public String getCategory() {
        return category;
    }

    public void setCategory(String category) {
        this.category = category;
    }

    public Integer getPId() {
        return id;
    }

    public void setId(Integer id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Float getPrice() {
        return price;
    }

    public void setPrice(Float price) {
        this.price = price;
    }

    public String getShipping() {
        return shipping;
    }

    public void setShipping(String shipping) {
        this.shipping = shipping;
    }


    @Override
    public String toString() {
        return "Product{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", price='" + price + '\'' +
                ", shipping=" + shipping +'\'' +
                ", image=" + image +'\'' +
                ", category=" + category +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Product product = (Product) o;
        return Objects.equals(getId(), product.getId()) &&
                Objects.equals(getName(), product.getName()) &&
                Objects.equals(getPrice(), product.getPrice()) &&
                Objects.equals(getShipping(), product.getShipping()) &&
                Objects.equals(getImage(), product.getImage()) &&
                Objects.equals(getCategory(), product.getCategory());
    }
    @Override
    public int hashCode() {
        return Objects.hash(getId(), getName(), getPrice(), getShipping(),getImage(),getCategory());
    }
}

File: cyberShop/src/main/frontend/src/app/product.ts
export class Product {
    id: number;
    name: String;
    category: String;
    shipping: String;
    price: number;
    tax: number;
    duty: number;
    quantity: number;
    available: number;
    image: String;

    constructor(name: String, category: String, shipping: String, price: number, quantity: number, tax?: number, duty?: number, available?: number, image?: String) {
        this.name = name;
        this.category = category;
        this.shipping = shipping;
        this.price = price;
        this.quantity = quantity;
        this.available = available;
        this.image = image;
        if (this.category == 'Music' || this.category == 'Luxury Items' || this.category == 'Clothes') {
            this.tax = .10;
        }

        if (this.category == 'Medical' || this.category == 'Book' || this.category == 'Food') {
            this.tax = 0;
        }
        if (this.shipping == 'imported') {
            this.duty = .05;
        }
        if (this.shipping == 'domestic') {
            this.duty = 0;
        }
    }
}
